### PR TITLE
Add core-test recording-playback

### PR DESCRIPTION
**Pull request by vhvb1989**
Adding recording on this PR.

Playback is not yet implemented.

### CODE REVIEW

File: sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace Azure { namespace Core { namespace Test {
+  class TestContextManager {
+  public:
+    std::string RecordingPath;
+    Azure::Core::Test::TestMode TestMode;
+    bool LiveOnly = false;
+
+    TestContextManager() { SetLiveOnly(); }
+
```
Comment (gearama) [inline sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp:23]: comments

File: sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * @return `true` when test is expected to be recorded or `false` when it is not.
+     */
+    bool DoNotRecordTest() const
+    {
+      return TestMode != Azure::Core::Test::TestMode::RECORD || LiveOnly;
+    }
+
+    bool IsPlaybackMode() const { return TestMode == Azure::Core::Test::TestMode::PLAYBACK; }
+    bool IsLiveMode() const { return TestMode == Azure::Core::Test::TestMode::LIVE; }
+
```
Comment (gearama) [inline sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp:69]: we could also add a isRecordMode() for completion

File: sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  protected:
+    constexpr static const char* LiveOnlyToken = "LIVE";
+
+  private:
+    std::string m_testName;
+    std::string m_testSuite;
+    void SetLiveOnly()
```
Comment (gearama) [inline sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp:76]: sould we also have a SetPlaybackRecordTest() method ?

File: sdk/core/azure-core-test/src/environment.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "private/environment.hpp"
+
```
Comment (gearama) [inline sdk/core/azure-core-test/src/environment.cpp:5]: weird empty line

File: sdk/core/azure-core-test/src/environment.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#endif
+#endif
+
+  return std::string();
+}
+
+Azure::Core::Test::TestMode Environment::GetTestMode()
+{
+  auto value = Environment::GetVariable("AZURE_TEST_MODE");
+  if (value.empty())
```
Comment (gearama) [inline sdk/core/azure-core-test/src/environment.cpp:51]: this should be the final else case on the if below

File: sdk/core/azure-core-test/src/environment.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  {
+    return Azure::Core::Test::TestMode::LIVE;
+  }
+
+  if (Azure::Core::_internal::StringExtensions::LocaleInvariantCaseInsensitiveEqual(
+          value, "RECORD"))
```
Comment (gearama) [inline sdk/core/azure-core-test/src/environment.cpp:57]: could we have these a string consts ?

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/recordings/KeyVaultClientTest.CreateKey123.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        "x-ms-keyvault-network-info": "conn_type=Ipv4;addr=20.49.4.206;act_addr_fam=InterNetwork;",
+        "x-ms-keyvault-region": "westus2",
+        "x-ms-keyvault-service-version": "1.9.48.0",
+        "x-ms-request-id": "29f865ca-f7bb-4d41-bea0-fbb17790785b",
+        "x-powered-by": "ASP.NET"
+      },
+      "Url": "https://REDACTED.vault.azure.net?api-version=7.2"
+    }
+  ]
+}
```
Comment (gearama) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/recordings/KeyVaultClientTest.CreateKey123.json:58]: newline at the end of recording

File: sdk/core/azure-core-test/src/test_base.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  }
+  root["networkCallRecords"] = records;
+
+  // Write json to file
+  std::ofstream outFile;
+  // AZURE_TEST_RECORDING_DIR is exported from CMAKE
+  std::string testPath(m_testContext.RecordingPath);
+  outFile.open(testPath + "/" + m_testContext.GetTestPlaybackRecordingName() + ".json");
+  outFile << root.dump(2);
+  outFile.close();
```
Comment (gearama) [inline sdk/core/azure-core-test/src/test_base.cpp:36]: outfile << "\n"

**Review by gearama — CHANGES_REQUESTED** [review]

File: sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp:23]: all this package is not expected to be published or exposed to customers. 
We won't generate docs to be published or so.

I am happy to add comments anyway for consistency, but the comments are just for ourselves

File: sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp:69]: Can be added later, sure.

File: sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp (hunk 3 above)
Comment (vhvb1989) [reply sdk/core/azure-core-test/inc/azure/core/test/test_context_manager.hpp:76]: We don't need to.

This `SetLiveOnly` is a special way how we can make some tests to ignore the RECORD or PLAYBACK configuration and always run LIVE.  
Some tests are hard to record. For example, streaming a blob. The pipeline returns a response with no payload at all and later the customer can start pulling data from network.  This is not supported yet. So, those tests need to be named as LIVE_ONLY and the recorder will ignore it.

File: sdk/core/azure-core-test/src/environment.cpp (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/azure-core-test/src/environment.cpp:51]: No need to check all options when value is empty.

File: sdk/core/azure-core-test/src/environment.cpp (hunk 6 above)
Comment (vhvb1989) [reply sdk/core/azure-core-test/src/environment.cpp:57]: I will check if it is used elsewhere. Otherwise, I don't see a real benefit

**Review by gearama — APPROVED** [review]

**Pull request merged**